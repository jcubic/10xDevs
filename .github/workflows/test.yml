name: 'Test'

on:
  push:
    branches: [master]

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '20'

      - name: Install dependencies
        run: npm ci

      - name: Run ESLint
        run: npm run lint

      - name: Run Prettier check
        run: npx prettier --check .

  unit-integration-tests:
    runs-on: ubuntu-latest
    needs: [lint]
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '20'

      - name: Install dependencies
        run: npm ci

      - name: Run unit/integration tests with coverage
        run: npm run test:coverage

      - name: Upload unit coverage artifact
        uses: actions/upload-artifact@v5
        with:
          name: unit-coverage
          path: coverage/unit/
          retention-days: 1

  e2e-tests:
    runs-on: ubuntu-latest
    needs: [lint]
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '20'

      - name: Install dependencies
        run: npm ci

      - name: Get Playwright version
        id: playwright-version
        run: echo "PLAYWRIGHT_VERSION=$(node -e "console.log(require('./package-lock.json').dependencies['@playwright/test'].version)")" >> $GITHUB_OUTPUT

      - name: Cache Playwright browsers
        uses: actions/cache@v4
        id: playwright-cache
        with:
          path: ~/.cache/ms-playwright
          key: ${{ runner.os }}-playwright-${{ steps.playwright-version.outputs.version }}

      - name: Install Playwright browsers
        run: npx playwright install --with-deps chromium
        if: steps.playwright-cache.outputs.cache-hit != 'true'

      - name: Install Playwright system dependencies
        run: npx playwright install-deps chromium
        if: steps.playwright-cache.outputs.cache-hit == 'true'

      - name: Set up E2E database and run tests
        id: e2e-tests
        run: |
          npm run test:e2e:coverage
          if [ -d e2e/results ] && [ "$(ls e2e/results)" ]; then
            echo "E2E test artifacts found, uploading..."
            echo "UPLOAD_ARTIFACTS=true" >> "$GITHUB_OUTPUT"
          else
            echo "No E2E test artifacts to upload"
            echo "UPLOAD_ARTIFACTS=false" >> "$GITHUB_OUTPUT"
          fi
        env:
          CI: 'true'
          COVERAGE: 'true'
          BASE_URL: 'http://localhost:45678'
          BETTER_AUTH_SECRET: 'test-secret-key-at-least-32-characters-long-for-ci'
          BETTER_AUTH_URL: 'http://localhost:45678'

      - name: Upload E2E artifacts to storage
        if: always() && steps.e2e-tests.outputs.UPLOAD_ARTIFACTS == 'true'
        uses: actions/upload-artifact@v5
        with:
          name: e2e-test-results
          path: e2e/results/
          retention-days: 7

      - name: Display E2E coverage report
        if: always()
        run: |
          if [ -d .nyc_output ] && [ "$(ls -A .nyc_output)" ]; then
            echo "E2E Coverage Report:"
            npx nyc report --reporter=text --temp-dir=.nyc_output
          else
            echo "No E2E coverage collected (requires webpack mode, not available in CI)"
          fi

      - name: Convert E2E coverage to lcov
        run: |
          mkdir -p coverage/e2e
          if [ -d .nyc_output ] && [ "$(ls -A .nyc_output)" ]; then
            echo "Converting E2E coverage to lcov format..."
            npx nyc report --reporter=lcov --report-dir=coverage/e2e --temp-dir=.nyc_output
            echo "E2E coverage converted"
          else
            echo "No E2E coverage found (webpack mode not used in CI)"
            touch coverage/e2e/lcov.info
          fi

      - name: Upload E2E coverage artifact
        uses: actions/upload-artifact@v5
        with:
          name: e2e-coverage
          path: coverage/e2e/
          retention-days: 1

  merge-coverage:
    runs-on: ubuntu-latest
    needs: [unit-integration-tests, e2e-tests]
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '20'

      - name: Install dependencies
        run: npm ci

      - name: Download unit coverage artifact
        uses: actions/download-artifact@v4
        with:
          name: unit-coverage
          path: coverage/unit/

      - name: Download E2E coverage artifact
        uses: actions/download-artifact@v4
        with:
          name: e2e-coverage
          path: coverage/e2e/

      - name: Merge coverage reports
        run: |
          mkdir -p coverage/merged

          # Ensure coverage files exist
          if [ ! -f coverage/unit/lcov.info ]; then
            echo "Warning: No unit coverage found"
            touch coverage/unit/lcov.info
          fi

          if [ ! -f coverage/e2e/lcov.info ]; then
            echo "Warning: No E2E coverage found"
            touch coverage/e2e/lcov.info
          fi

          # Merge coverage files by concatenation
          echo "Merging coverage reports..."
          cat coverage/unit/lcov.info coverage/e2e/lcov.info > coverage/merged/lcov.info

          echo "Coverage merge complete"

      - name: Display merged coverage summary
        if: always()
        run: |
          if [ -f coverage/merged/lcov.info ] && [ -s coverage/merged/lcov.info ]; then
            echo "Merged Coverage Summary:"
            npx lcov-summary coverage/merged/lcov.info
          else
            echo "No merged coverage summary available"
          fi

      - name: Upload merged coverage to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          path-to-lcov: ./coverage/merged/lcov.info
