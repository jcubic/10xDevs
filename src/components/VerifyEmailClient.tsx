'use client';

import { useRouter } from 'next/navigation';
import { Box, Heading, VStack, Text, Alert, Button } from '@chakra-ui/react';
import Link from 'next/link';
import { useCountdown } from '@/hooks/useCountdown';

type VerifyEmailClientProps = {
  success: boolean;
  error?: string;
  isExpired?: boolean;
};

export default function VerifyEmailClient({
  success,
  error,
  isExpired
}: VerifyEmailClientProps) {
  const router = useRouter();

  const countdown = useCountdown({
    initialCount: 3,
    onComplete: () => router.push('/'),
    enabled: success
  });

  return (
    <Box
      p={6}
      maxW="md"
      mx="auto"
      bg="bg"
      minH="100vh"
      display="flex"
      alignItems="center"
      justifyContent="center"
    >
      <VStack gap={6} align="stretch" width="full">
        {success ? (
          <>
            <Heading size="xl" color="fg" textAlign="center">
              Email Verified!
            </Heading>

            <Alert.Root status="success">
              <Alert.Indicator />
              <Alert.Content>
                <Alert.Title>Welcome to SNApp!</Alert.Title>
                <Alert.Description>
                  Your email has been successfully verified. You can now access all
                  features.
                </Alert.Description>
              </Alert.Content>
            </Alert.Root>

            <VStack gap={4} align="stretch">
              <Text textAlign="center" color="fg.muted" fontSize="sm">
                Redirecting to the app in {countdown} seconds...
              </Text>

              <Button onClick={() => router.push('/')} size="lg">
                Go to App
              </Button>

              <Text textAlign="center" color="fg.muted" fontSize="sm">
                Or{' '}
                <Link href="/login" style={{ color: 'var(--chakra-colors-blue-500)' }}>
                  sign in here
                </Link>
              </Text>
            </VStack>
          </>
        ) : (
          <>
            <Heading size="xl" color="fg" textAlign="center">
              Verification Failed
            </Heading>

            <Alert.Root status="error">
              <Alert.Indicator />
              <Alert.Content>
                <Alert.Title>
                  {isExpired ? 'Link Expired' : 'Verification Error'}
                </Alert.Title>
                <Alert.Description>{error}</Alert.Description>
              </Alert.Content>
            </Alert.Root>

            <VStack gap={4} align="stretch">
              {isExpired && (
                <Text textAlign="center" color="fg.muted" fontSize="sm">
                  Verification links expire after 24 hours for security reasons.
                </Text>
              )}

              <Button
                onClick={() => router.push('/register')}
                variant="outline"
                size="lg"
              >
                {isExpired ? 'Register Again' : 'Try Again'}
              </Button>

              <Text textAlign="center" color="fg.muted" fontSize="sm">
                Need help?{' '}
                <Link href="/login" style={{ color: 'var(--chakra-colors-blue-500)' }}>
                  Contact Support
                </Link>
              </Text>
            </VStack>
          </>
        )}
      </VStack>
    </Box>
  );
}
