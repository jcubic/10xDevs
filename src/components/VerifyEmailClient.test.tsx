import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen, act } from '@/test/utils';
import VerifyEmailClient from './VerifyEmailClient';

// Mock next/navigation
const mockPush = vi.fn();
vi.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush
  })
}));

describe('VerifyEmailClient', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe('Success State', () => {
    it('matches snapshot for successful verification', () => {
      const { container } = render(<VerifyEmailClient success={true} />);
      expect(container).toMatchSnapshot();
    });

    it('displays success message and countdown', () => {
      render(<VerifyEmailClient success={true} />);

      expect(screen.getByText('Email Verified!')).toBeInTheDocument();
      expect(screen.getByText('Welcome to SNApp!')).toBeInTheDocument();
      expect(
        screen.getByText(
          'Your email has been successfully verified. You can now access all features.'
        )
      ).toBeInTheDocument();
      expect(screen.getByText(/Redirecting to the app in 3 seconds/)).toBeInTheDocument();
    });

    it('updates countdown from 3 to 0', () => {
      render(<VerifyEmailClient success={true} />);

      expect(screen.getByText(/Redirecting to the app in 3 seconds/)).toBeInTheDocument();

      act(() => {
        vi.advanceTimersByTime(1000);
      });
      expect(screen.getByText(/Redirecting to the app in 2 seconds/)).toBeInTheDocument();

      act(() => {
        vi.advanceTimersByTime(1000);
      });
      expect(screen.getByText(/Redirecting to the app in 1 seconds/)).toBeInTheDocument();
    });

    it('redirects to home page after countdown completes', () => {
      render(<VerifyEmailClient success={true} />);

      act(() => {
        vi.advanceTimersByTime(3000);
      });

      expect(mockPush).toHaveBeenCalledWith('/');
    });

    it('redirects immediately when "Go to App" button is clicked', async () => {
      vi.useRealTimers();
      const { user } = render(<VerifyEmailClient success={true} />);

      const goToAppButton = screen.getByRole('button', { name: /go to app/i });
      await user.click(goToAppButton);

      expect(mockPush).toHaveBeenCalledWith('/');
    });

    it('displays link to sign in page', () => {
      render(<VerifyEmailClient success={true} />);

      const signInLink = screen.getByRole('link', { name: /sign in here/i });
      expect(signInLink).toHaveAttribute('href', '/login');
    });
  });

  describe('Error State - Generic Error', () => {
    it('matches snapshot for generic error', () => {
      const { container } = render(
        <VerifyEmailClient
          success={false}
          error="Email verification failed"
          isExpired={false}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('displays generic error message', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="Email verification failed"
          isExpired={false}
        />
      );

      expect(screen.getByText('Verification Failed')).toBeInTheDocument();
      expect(screen.getByText('Verification Error')).toBeInTheDocument();
      expect(screen.getByText('Email verification failed')).toBeInTheDocument();
    });

    it('displays "Try Again" button for generic error', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="Email verification failed"
          isExpired={false}
        />
      );

      const tryAgainButton = screen.getByRole('button', { name: /try again/i });
      expect(tryAgainButton).toBeInTheDocument();
    });

    it('does not display expiry message for generic error', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="Email verification failed"
          isExpired={false}
        />
      );

      expect(
        screen.queryByText(/verification links expire after 24 hours/i)
      ).not.toBeInTheDocument();
    });
  });

  describe('Error State - Expired Link', () => {
    it('matches snapshot for expired link', () => {
      const { container } = render(
        <VerifyEmailClient
          success={false}
          error="This verification link has expired or is invalid"
          isExpired={true}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('displays expired link error message', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="This verification link has expired or is invalid"
          isExpired={true}
        />
      );

      expect(screen.getByText('Verification Failed')).toBeInTheDocument();
      expect(screen.getByText('Link Expired')).toBeInTheDocument();
      expect(
        screen.getByText('This verification link has expired or is invalid')
      ).toBeInTheDocument();
    });

    it('displays "Register Again" button for expired link', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="This verification link has expired or is invalid"
          isExpired={true}
        />
      );

      const registerAgainButton = screen.getByRole('button', { name: /register again/i });
      expect(registerAgainButton).toBeInTheDocument();
    });

    it('displays expiry explanation message', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="This verification link has expired or is invalid"
          isExpired={true}
        />
      );

      expect(
        screen.getByText('Verification links expire after 24 hours for security reasons.')
      ).toBeInTheDocument();
    });

    it('redirects to register page when "Register Again" is clicked', async () => {
      vi.useRealTimers();
      const { user } = render(
        <VerifyEmailClient
          success={false}
          error="This verification link has expired or is invalid"
          isExpired={true}
        />
      );

      const registerAgainButton = screen.getByRole('button', { name: /register again/i });
      await user.click(registerAgainButton);

      expect(mockPush).toHaveBeenCalledWith('/register');
    });
  });

  describe('Common Elements', () => {
    it('displays contact support link in error states', () => {
      render(
        <VerifyEmailClient
          success={false}
          error="Email verification failed"
          isExpired={false}
        />
      );

      const supportLink = screen.getByRole('link', { name: /contact support/i });
      expect(supportLink).toHaveAttribute('href', '/login');
    });

    it('cleans up countdown timer on unmount', () => {
      const { unmount } = render(<VerifyEmailClient success={true} />);

      // Start countdown
      act(() => {
        vi.advanceTimersByTime(1000);
      });
      expect(screen.getByText(/Redirecting to the app in 2 seconds/)).toBeInTheDocument();

      // Unmount component
      unmount();

      // Advance timers - should not cause any errors
      act(() => {
        vi.advanceTimersByTime(3000);
      });

      // No router push should happen after unmount
      expect(mockPush).not.toHaveBeenCalled();
    });
  });
});
