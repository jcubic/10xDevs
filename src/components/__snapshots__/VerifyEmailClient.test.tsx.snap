// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VerifyEmailClient > Error State - Expired Link > matches snapshot for expired link 1`] = `
<div>
  <div
    class="css-qzj89r"
  >
    <div
      class="chakra-stack css-g3f1vx"
    >
      <h2
        class="chakra-heading css-15pb32e"
      >
        Verification Failed
      </h2>
      <div
        class="chakra-alert__root css-1kf3jln"
      >
        <span
          class="css-13gnd3a"
        >
          <svg
            class="css-126y6n1"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11 15H13V17H11V15ZM11 7H13V13H11V7Z"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__content css-xlayl9"
        >
          <div
            class="chakra-alert__title css-1kctzi1"
          >
            Link Expired
          </div>
          <div
            class="chakra-alert__description css-t6kp1r"
          >
            This verification link has expired or is invalid
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-18abhre"
      >
        <p
          class="css-j7r03v"
        >
          Verification links expire after 24 hours for security reasons.
        </p>
        <button
          class="chakra-button css-5yfs8l"
          type="button"
        >
          Register Again
        </button>
        <p
          class="css-j7r03v"
        >
          Need help?
           
          <a
            href="/login"
            style="color: var(--chakra-colors-blue-500);"
          >
            Contact Support
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`VerifyEmailClient > Error State - Generic Error > matches snapshot for generic error 1`] = `
<div>
  <div
    class="css-qzj89r"
  >
    <div
      class="chakra-stack css-g3f1vx"
    >
      <h2
        class="chakra-heading css-15pb32e"
      >
        Verification Failed
      </h2>
      <div
        class="chakra-alert__root css-1kf3jln"
      >
        <span
          class="css-13gnd3a"
        >
          <svg
            class="css-126y6n1"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11 15H13V17H11V15ZM11 7H13V13H11V7Z"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__content css-xlayl9"
        >
          <div
            class="chakra-alert__title css-1kctzi1"
          >
            Verification Error
          </div>
          <div
            class="chakra-alert__description css-t6kp1r"
          >
            Email verification failed
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-18abhre"
      >
        <button
          class="chakra-button css-5yfs8l"
          type="button"
        >
          Try Again
        </button>
        <p
          class="css-j7r03v"
        >
          Need help?
           
          <a
            href="/login"
            style="color: var(--chakra-colors-blue-500);"
          >
            Contact Support
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`VerifyEmailClient > Success State > matches snapshot for successful verification 1`] = `
<div>
  <div
    class="css-qzj89r"
  >
    <div
      class="chakra-stack css-g3f1vx"
    >
      <h2
        class="chakra-heading css-15pb32e"
      >
        Email Verified!
      </h2>
      <div
        class="chakra-alert__root css-1c44bi3"
      >
        <span
          class="css-13gnd3a"
        >
          <svg
            class="css-126y6n1"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11.0026 16L6.75999 11.7574L8.17421 10.3431L11.0026 13.1716L16.6595 7.51472L18.0737 8.92893L11.0026 16Z"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__content css-xlayl9"
        >
          <div
            class="chakra-alert__title css-1kctzi1"
          >
            Welcome to SNApp!
          </div>
          <div
            class="chakra-alert__description css-t6kp1r"
          >
            Your email has been successfully verified. You can now access all features.
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-18abhre"
      >
        <p
          class="css-j7r03v"
        >
          Redirecting to the app in 
          3
           seconds...
        </p>
        <button
          class="chakra-button css-1u8nmh6"
          type="button"
        >
          Go to App
        </button>
        <p
          class="css-j7r03v"
        >
          Or
           
          <a
            href="/login"
            style="color: var(--chakra-colors-blue-500);"
          >
            sign in here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
